include/session.hh

```diff
@@ -27,63 +27,104 @@ public:
 
     ~session();
 
+    prepared_query prepare(const std::string &query);
+
+    const CassPrepared *prepare_raw(const std::string& query);
+
     future execute_async(const statement &statement);
+    future execute_async(const statement &&statement);
 
     // Convenience functions to avoid creating scd_statement manually for simple queries
     future execute_async(const std::string &query);
+    future execute_async(const std::string &&query);
 
-    future execute_async(const batch_query& query);
+    future execute_async(const batch_query &query);
+    future execute_async(const batch_query &&query);
 
-    future execute_async(const prepared_query& query);
+    future execute_async(const prepared_query &query);
+    future execute_async(const prepared_query &&query);
 
     template<typename... Args,
         typename = typename std::enable_if<0 != sizeof...(Args)>::type>
     future execute_async(const std::string &query, Args... args) {
         return execute_async(scmd::statement(query, sizeof...(args)).bind(args...));
     }
+    template<typename... Args,
+        typename = typename std::enable_if<0 != sizeof...(Args)>::type>
+    future execute_async(const std::string &&query, Args... args) {
+        return execute_async(scmd::statement(query, sizeof...(args)).bind(args...));
+    }
+
 
     template<typename... Args,
         typename = typename std::enable_if<0 != sizeof...(Args)>::type>
     future execute_async(statement &statement, Args... args) {
         return execute_async(statement.bind(args...));
     }
-
     template<typename... Args,
         typename = typename std::enable_if<0 != sizeof...(Args)>::type>
-    future execute_async(const prepared_query& query, Args... args) {
-        auto statement = query.get_statement();
+    future execute_async(const statement &&statement, Args... args) {
         return execute_async(statement.bind(args...));
```

Comment:
Not sure of how moving works here (what does 'bind' return? a value? or a reference? if it's a reference, is this what we really intend to do? well, perhaps yes, but still, careful with this).